### ISSUE TITLE
Hash the primary inputs to save a few scalar mult in the verifier circuit

### ISSUE THREAD

**Issue by AntoineRondelet** [clearmatics/zecale#5]

**Comment by AntoineRondelet** [comment]
This is necessary to make Zecale generic.

**Comment by dtebbs** [comment]
Note, it may not be feasible to do this via an algebraic hash (without support for some extra precompiled contracts) since scalar field elements do not fit into single evm words.  In the case of verification keys, hashing with Blake2 increased the proving time of the circuit for (2 nested proofs) by an roughly order of magnitude.

**Comment by AntoineRondelet** [comment]
> Note, it may not be feasible to do this via an algebraic hash (without support for some extra precompiled contracts) since scalar field elements do not fit into single evm words. In the case of verification keys, hashing with Blake2 increased the proving time of the circuit for (2 nested proofs) by an roughly order of magnitude.

Yes, we need to make the assumption that the EVM is extended with another set of precompiled contracts (even in its current state, Zecale cannot be deployed on Ethereum mainnet because BW6 arithmetic is not supported)
